### PR TITLE
Rename dns_cache_max_size to dns_cache_max_entries

### PR DESCRIPTION
**Pull request by allmazz**
See https://github.com/ClickHouse/ClickHouse/pull/60257#discussion_r1506200824

### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Renamed server setting `dns_cache_max_size` to `dns_cache_max_entries` to reduce ambiguity

### CODE REVIEW

**Comment by pufit** [comment]
The CacheBase class here https://github.com/ClickHouse/ClickHouse/blob/master/src/Common/DNSResolver.cpp#L179 takes an optional template argument `WeightFunction`.

The default weight function is this
https://github.com/ClickHouse/ClickHouse/blob/master/src/Common/ICachePolicy.h#L15

So we need to change either the setting (e.g., `dns_cache_max_entries`) or add a correct weight function.

File: src/Core/ServerSettings.h [hunk 1]
```diff
@@ -81,7 +81,7 @@ namespace DB
     M(UInt64, mmap_cache_size, DEFAULT_MMAP_CACHE_MAX_SIZE, "A cache for mmapped files.", 0) \
     \
     M(Bool,   disable_internal_dns_cache, false, "Disable internal DNS caching at all.", 0) \
-    M(UInt64, dns_cache_max_size, 1024, "Internal DNS cache max size in bytes.", 0) \
+    M(UInt64, dns_cache_max_size, (1024 * 1024), "Internal DNS cache max size in bytes.", 0) \
```
Comment (rschu1ze) [inline src/Core/ServerSettings.h:84]: We should `#include <base/unit.h>` and use more readable quantities like `1_MB`.

**Comment by rschu1ze** [comment]
> The CacheBase class here https://github.com/ClickHouse/ClickHouse/blob/master/src/Common/DNSResolver.cpp#L179 takes an optional template argument `WeightFunction`.
> 
> The default weight function is this https://github.com/ClickHouse/ClickHouse/blob/master/src/Common/ICachePolicy.h#L15
> 
> So we need to change either the setting (e.g., `dns_cache_max_entries`) or add a correct weight function.

Just curious: why? The purpose of this PR is to increase the DNS cache size, and not change the relative weight of entries.

**Comment by pufit** [comment]
> Just curious: why? The purpose of this PR is to increase the DNS cache size (see [#60257 (comment)](https://github.com/ClickHouse/ClickHouse/pull/60257#discussion_r1506200824)), and not change the relative weight of entries.

IMHO, `dns_cache_max_size` should limit the actual size of the cache in bytes. Otherwise, the setting's name is very counterintuitive.

**Comment by rschu1ze** [comment]
I agree that the naming is ambiguous. It is not clear if it means the number of cache entries of the total cache size in bytes.

For reference, I had a similar naming challenge when building the query cache a while ago. I solved the naming like that:

```cpp
/// from server/Server.cpp
size_t query_cache_max_size_in_bytes = config().getUInt64("query_cache.max_size_in_bytes", DEFAULT_QUERY_CACHE_MAX_SIZE);
size_t query_cache_max_entries = config().getUInt64("query_cache.max_entries", DEFAULT_QUERY_CACHE_MAX_ENTRIES);
size_t query_cache_query_cache_max_entry_size_in_bytes = config().getUInt64("query_cache.max_entry_size_in_bytes", DEFAULT_QUERY_CACHE_MAX_ENTRY_SIZE_IN_BYTES);
size_t query_cache_max_entry_size_in_rows = config().getUInt64("query_cache.max_entry_rows_in_rows", DEFAULT_QUERY_CACHE_MAX_ENTRY_SIZE_IN_ROWS);
```

The setting name was introduced [only a week ago](https://github.com/ClickHouse/ClickHouse/pull/60257), we still have a chance to change the naming until a new release is out (*).

(*) The sad news is that the new release v24.2 will likely be made tomorrow in the evening, CET. But we could backport this PR as I don't expect that the majority of users will mess with the new setting.

File: src/Common/DNSResolver.cpp [hunk 2]
```diff
@@ -289,10 +289,10 @@ void DNSResolver::setDisableCacheFlag(bool is_disabled)
     impl->disable_cache = is_disabled;
 }
 
-void DNSResolver::setCacheMaxSize(const UInt64 cache_max_size)
+void DNSResolver::setCacheMaxEntries(const UInt64 cache_max_entries)
 {
-    impl->cache_address.setMaxSizeInBytes(cache_max_size);
-    impl->cache_host.setMaxSizeInBytes(cache_max_size);
+    impl->cache_address.setMaxSizeInBytes(cache_max_entries);
```
Comment (rschu1ze) [inline src/Common/DNSResolver.cpp:294]: Wait a second. Maybe there is a misunderstanding. `setMaxSizeInBytes` sets the maximum total size of the cache. If you like to set the maximum total number of entries, you need to use `setMaxCount`.

File: docs/en/operations/server-configuration-parameters/settings.md [hunk 3]
```diff
@@ -200,17 +200,17 @@ Type: Bool
 Default: 0
 
 
-## dns_cache_max_size
+## dns_cache_max_entries
 
-Internal DNS cache max size in bytes.
+Internal DNS cache max entries.
```
Comment (rschu1ze) [inline docs/en/operations/server-configuration-parameters/settings.md:205]: Suggest to remove the comment in l. 207-209. It talks about the memory size in bytes but this setting is about the number of entries.

File: src/Common/DNSResolver.cpp (hunk 2 above)
Comment (allmazz) [reply src/Common/DNSResolver.cpp:294]: See discussion above. Default weight policy return 1 on any entry, so setMaxSizeInBytes actually is count

File: docs/en/operations/server-configuration-parameters/settings.md (hunk 3 above)
Comment (allmazz) [reply docs/en/operations/server-configuration-parameters/settings.md:205]: yes, i removed it

**Review by rschu1ze — APPROVED** [review]

**Comment by robot-ch-test-poll1** [comment]
*This is an automated comment for commit be552076997eac4a467668748684c8dc8df82e10 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/60500/be552076997eac4a467668748684c8dc8df82e10/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

**Pull request merged**